sdk/sdk/src/main/java/org/apache/carbondata/sdk/file/PaginationCarbonReader.java

```diff
@@ -0,0 +1,303 @@
+
+  private Range getBlockletIndexRange(long fromRowNumber, long toRowNumber) {
+    // find the matching blocklet index range by checking with from and to.
+    int upperBound = findBlockletIndex(rowCountInSplits, toRowNumber);
+    // lower bound cannot be more than upper bound, so work on sub list.
+    int lowerBound = findBlockletIndex(rowCountInSplits.subList(0, upperBound), fromRowNumber);
+    return new Range(lowerBound, upperBound);
+  }
+
+  private Object[] getRows(long fromRowNumber, long toRowNumber)
+      throws IOException, InterruptedException {
+    int rowCount = 0;
+    Object[] rows = new Object[(int)(toRowNumber - fromRowNumber + 1)];
+    // get the matching split index (blocklets) range for the input range.
+    Range blockletIndexRange = getBlockletIndexRange(fromRowNumber, toRowNumber);
+    for (int i = blockletIndexRange.getFrom(); i <= blockletIndexRange.getTo(); i++) {
+      String blockletUniqueId = String.valueOf(i);
+      BlockletRows blockletRows;
+      if (cache.get(blockletUniqueId) != null) {
+        blockletRows = (BlockletRows)cache.get(blockletUniqueId);
+      } else {
+        BlockletDetailInfo detailInfo =
+            ((CarbonInputSplit) allBlockletSplits.get(i)).getDetailInfo();
+        int rowCountInBlocklet = detailInfo.getRowCount();
+        Object[] rowsInBlocklet = new Object[rowCountInBlocklet];
+        // read the rows from the blocklet
+        // TODO: read blocklets in multi-thread if there is a performance requirement.
+        readerBuilder.setInputSplit(allBlockletSplits.get(i));
+        CarbonReader<Object> carbonReader = readerBuilder.build();
+        int count = 0;
+        while (carbonReader.hasNext()) {
+          rowsInBlocklet[count++] = carbonReader.readNextRow();
+        }
+        carbonReader.close();
+        long fromRowId;
+        if (i == 0) {
+          fromRowId = 1;
+        } else {
+          // previous index will contain the sum of rows till previous blocklet.
+          fromRowId = rowCountInSplits.get(i - 1) + 1;
+        }
+        blockletRows = new BlockletRows(fromRowId, detailInfo.getBlockSize(), rowsInBlocklet);
+        // add entry to cache with no expiry time
+        // key: unique blocklet id
+        // value: BlockletRows
+        cache.put(String.valueOf(i), blockletRows, blockletRows.getMemorySize(), Integer.MAX_VALUE);
+      }
+      long fromBlockletRow = blockletRows.getRowIdStartIndex();
+      long toBlockletRow = fromBlockletRow + blockletRows.getRowsCount();
+      Object[] rowsInBlocklet = blockletRows.getRows();
+      if (toRowNumber > toBlockletRow) {
+        if (fromRowNumber >= fromBlockletRow) {
+          // only fromRowNumber lies in this blocklet,
+          // read from fromRowNumber to end of the blocklet.
+          // -1 because row id starts form 0
+          int start = (int) (fromRowNumber - blockletRows.getRowIdStartIndex());
+          int end = blockletRows.getRowsCount();
+          while (start < end) {
```

Comment:
All 3 partial overlap cases we copy selectively. so start and end changes. but the while loop copy is same. How about refactor such that copy is moved out to common ? Just a suggestion if it doesn't make code difficult to understand.

Reply:
it will be hard to read. I feel current changes is OK.